### PR TITLE
[5.4] Extending Blade's looping functionality

### PR DESCRIPTION
**Pull request by themsaid**
As noted in this issue https://github.com/laravel/framework/issues/12601, this PR adds extra functionality for Blade's loops:

``` php
@foreach($items as $item)
     {{ $loop->index }} // The current iteration starting 1.
     {{ $loop->reverseIndex }} // The number of iterations left till the end. Ends with 1.
     {{ $loop->size }} // The size of the iteratee.
     {{ $loop->isFirst }} // If the current iteration is the first. (boolean)
     {{ $loop->isLast }} // If the current iteration is the last. (boolean)
     {{ $loop->depth }} // The depth inside a nested loop.
     {{ $loop->parent }} // The direct parent loop information.

    {{ $loop->parent->isFirst }} // This checks if the parent loop is at the first iteration.
@endforeach
```

`$loop` carries information about the current iteration of the loop it was called inside, outside loops `$loop` will be `null`.

---

**Note:** This proposal is applied only on `@foreach` for now but I'm willing to apply it on `@for` & `@forelse` as well if approved and after iterations.

### CODE REVIEW

File: src/Illuminate/Foundation/helpers.php [hunk 1]
```diff
@@ -780,3 +780,17 @@ function view($view = null, $data = [], $mergeData = [])
         return $factory->make($view, $data, $mergeData);
     }
 }
+
+if (! function_exists('loop')) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: No way we should be defining foundation helpers.

Comment (themsaid) [inline src/Illuminate/Foundation/helpers.php]: Moved to support helpers.

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: that's worse!

Comment (themsaid) [inline src/Illuminate/Foundation/helpers.php]: Then what do you suggest?

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: not using helpers

Comment (themsaid) [inline src/Illuminate/Foundation/helpers.php]: I thought of having a blade directive `@loop('isFirst')` `@loop('isLast')` but then it won't be available inside `<?php if(@loop('isLast'))) //; ?>` code.

**Comment by mark86092** [comment]
If someone use `@continue` inside, `$__env->incrementLoopIndices();` is skipped, causing problem of index counting

**Comment by themsaid** [comment]
@mark86092 you are right, the call to the increment method should be at the beginning of the iteration.

File: src/Illuminate/Support/helpers.php [hunk 2]
```diff
@@ -863,3 +863,17 @@ function with($object)
         return $object;
     }
 }
+
+if (! function_exists('loop')) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/helpers.php]: This can't go in here because it depends on components outside the support component.

Comment (themsaid) [inline src/Illuminate/Support/helpers.php]: I initially added it to the foundation helpers just like `view()` but that wasn't acceptable as well, the options are:

1. Add a helper somewhere (foundation helpers, support helpers)
2. A method in the View facade `View::loop()`.
3. Unique shared variable in views `$__loop`.

What do you think? are there any other options?

**Comment by Yannik** [comment]
Thanks for working on this, @themsaid!

File: src/Illuminate/View/Compilers/BladeCompiler.php [hunk 3]
```diff
@@ -579,7 +579,11 @@ protected function compileFor($expression)
      */
     protected function compileForeach($expression)
     {
-        return "<?php foreach{$expression}: ?>";
+        preg_match('/\( *(.*) *as /', $expression, $matches);
+
+        $content = trim($matches[1]);
+
+        return "<?php \$__env->addLoop(count({$content})); foreach{$expression}: \$__env->incrementLoopIndices(); ?>";
```
Comment (Yannik) [inline src/Illuminate/View/Compilers/BladeCompiler.php]: You should only `count` the object if it is `Countable`. In the case that it's not countable, revIndex/isLast/size will not be available, but the remaining functionality (isFirst, index) can still be used.

File: src/Illuminate/View/Factory.php [hunk 4]
```diff
@@ -693,6 +700,63 @@ public function doneRendering()
...
+     *
+     * @return void
+     */
+    public function incrementLoopIndices()
+    {
+        $loop = &$this->loopsStack[count($this->loopsStack) - 1];
+
+        $loop['index'] += 1;
+
+        $loop['reverseIndex'] -= 1;
```
Comment (Yannik) [inline src/Illuminate/View/Factory.php]: Only do this if `size` is set. (Which it should only be if the item being looped over is either an array or a object that is `Countable`)

File: src/Illuminate/View/Factory.php [hunk 5]
```diff
@@ -693,6 +700,63 @@ public function doneRendering()
...
+
+        $loop['isFirst'] = $loop['index'] == 1;
+
+        $loop['isLast'] = $loop['index'] == $loop['size'];
```
Comment (Yannik) [inline src/Illuminate/View/Factory.php]: Only do this if `size` is set. (Which it should only be if the item being looped over is either an array or a object that is `Countable`)

File: src/Illuminate/View/Factory.php [hunk 6]
```diff
@@ -693,6 +700,63 @@ public function doneRendering()
...
+    public function addLoop($length)
+    {
+        $this->loopsStack[] = [
+            'index' => 0,
+            'reverseIndex' => $length + 1,
+            'size' => $length,
+            'isFirst' => true,
+            'isLast' => $length == 1,
+            'depth' => count($this->loopsStack) + 1,
+            'parent' => count($this->loopsStack) ? Arr::last($this->loopsStack) : null,
```
Comment (mark86092) [inline src/Illuminate/View/Factory.php]: Arr::last($this->loopsStack) will return default null, it is OK that
```
'parent' => Arr::last($this->loopsStack)
```

**Comment by themsaid** [comment]
@Yannik & @mark86092 thank you for your notes, I adjusted the code so that if the value is not countable `reverseIndex`, `size`, & `isLast` will always be `null`.

File: src/Illuminate/View/Factory.php [hunk 7]
```diff
@@ -693,6 +700,67 @@ public function doneRendering()
...
+    public function addLoop($data)
+    {
+        $length = is_array($data) || $data instanceof \Countable ? count($data) : null;
+
+        $this->loopsStack[] = [
+            'index' => 0,
+            'reverseIndex' => isset($length) ? $length + 1 : null,
+            'size' => $length,
+            'isFirst' => true,
+            'isLast' => isset($length) ? $length == 1 : null,
```
Comment (Yannik) [inline src/Illuminate/View/Factory.php]: Should probably be false instead of null.

Comment (themsaid) [inline src/Illuminate/View/Factory.php]: But false means it's not the last one, while it can be the last one but we don't know. So maybe null is better.

Comment (Yannik) [inline src/Illuminate/View/Factory.php]: You are right!

**Comment by taylorotwell** [comment]
Personally I would rather just have a `$loop` variable than a `loop` global function.

**Comment by themsaid** [comment]
@taylorotwell code updated .. Now a `$loop` shared variable is shared across views and holds information about the current loop.

File: src/Illuminate/Support/helpers.php [hunk 8]
```diff
@@ -841,7 +841,7 @@ function value($value)
 
 if (! function_exists('windows_os')) {
     /**
-     * Determine whether the current envrionment is Windows based.
+     * Determine whether the current environment is Windows based.
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/helpers.php]: This fix is unrelated and should go into 5.2.

Comment (themsaid) [inline src/Illuminate/Support/helpers.php]: Removed.

**Comment by GrahamCampbell** [comment]
Please squash to one commit. ;)

**Comment by JoostK** [comment]
This evaluates the iteratee multiple times, which is very much not desirable.

**Comment by themsaid** [comment]
@GrahamCampbell  -----)squashed(-----

File: src/Illuminate/View/Factory.php [hunk 9]
```diff
@@ -113,6 +120,7 @@ public function __construct(EngineResolver $engines, ViewFinderInterface $finder
         $this->engines = $engines;
 
         $this->share('__env', $this);
+        $this->share('loop', (object) Arr::last($this->getLoopsStack()));
```
Comment (mark86092) [inline src/Illuminate/View/Factory.php]: The $loop is fixed after construction, how come the $loop is updated in each iteration?

File: tests/View/ViewFactoryTest.php [hunk 10]
```diff
@@ -427,6 +427,101 @@ public function testExtraAppendSectionCallThrowsException()
...
+        $factory->addLoop([1, 2, 3, 4]);
+
+        $factory->incrementLoopIndices();
+
+        $factory->incrementLoopIndices();
+
+        $this->assertEquals(2, $factory->getLoopsStack()[0]['index']);
+        $this->assertEquals(3, $factory->getLoopsStack()[0]['reverseIndex']);
+    }
+
```
Comment (mark86092) [inline tests/View/ViewFactoryTest.php]: I got an idea that you can test `$loop` in view by assert `$factory->shared('loop')`, for example
```
$this->assertEquals(2, $factory->shared('loop')->index)
```

**Comment by themsaid** [comment]
@mark86092 updating the shared variable is now taken care of, It'll get updates on every new loop, end of loop, or iteration.

**Comment by JoostK** [comment]
Now it totally breaks nested loops!

**Comment by themsaid** [comment]
@JoostK ok I'm going to tag [WIP] on this PR for now, it looks like doing rapid iterations caused a lot of garbage, I'm going to test it carefully and hint Taylor when I'm done.

Thanks for giving me hints @JoostK

**Comment by themsaid** [comment]
Testing the following view:
```
@foreach([1,2,3] as $i)
       1: {{$loop->index}}

        @foreach([1,2] as $i)
            2:{{$loop->index}}
        @endforeach

       1: {{$loop->index}}
@endforeach
```

Produces:

```
<?php $__currentLoopData = [1,2,3]; $__env->addLoop($__currentLoopData); foreach($__currentLoopData as $i): $__env->incrementLoopIndices(); $loop = $__env->getFirstLoop(); ?>
    1:<?php echo e($loop->index); ?>

    <?php $__currentLoopData = [1,2]; $__env->addLoop($__currentLoopData); foreach($__currentLoopData as $i): $__env->incrementLoopIndices(); $loop = $__env->getFirstLoop(); ?>
     2:<?php echo e($loop->index); ?>

    <?php endforeach; $__env->popLoop(); $loop = $__env->getFirstLoop(); ?>

   1: <?php echo e($loop->index); ?>
<?php endforeach; $__env->popLoop(); $loop = $__env->getFirstLoop(); ?>
```

And outputs the following:

```
1:1
2:1
2:2
1:1

1:2
2:1
2:2
1:2

1:3
2:1
2:2
1:3
```

**Comment by alexbowers** [comment]
For some added fun, not sure how feasible this is, and it really _isn't_ part of built in PHP, but a way to do `use ($loop as $parent_loop)` on inner loops would be good to get information from outer loops.

**Comment by alexbowers** [comment]
Or should that be perhaps a different PR?

**Comment by themsaid** [comment]
@alexbowers you may do that: 

```
$loop->parent->parent->index;
```

**Comment by alexbowers** [comment]
Ah cool, didn't know that'd work. Thats cool I guess :)

**Comment by JoostK** [comment]
I was actually thinking about another issue, but upon further thinking about it I was wrong. Sorry about that. Can you verify though that your outer loop `$i` works as intended (for which you'd have to rename the iterator variable in the inner loop, or it would override the outer variable)

**Comment by themsaid** [comment]
@JoostK Actually no, it'll override the variable of the parent loop, I just wasn't concerned about that while testing.

**Comment by alexbowers** [comment]
What would the effect of this be:

```php
@foreach ($iterator as $loop)
 {{ $loop }}
@endforeach
```

**Comment by arrilot** [comment]
I'm concerned about performance implications. We need to make sure it does slow done blade loops a lot

**Comment by Yannik** [comment]
The performance implications are likely near nil.

**Comment by mark86092** [comment]
> What would the effect of this be:
> ```
> @foreach ($iterator as $loop)
>  {{ $loop }}
> @endforeach
> ```

maybe `$__loop = $__env->getFirstLoop();` instead ?

**Comment by Yannik** [comment]
@mark86092 Pretty ugly. As this is proposed for 5.3, a bit of backwards-breaking changes are to be expected IMO.

**Comment by sileence** [comment]
I'm late to the discussion, but what if we could have this:

`@foreach($items as $item with $loop)` 

or this

`@foreach($items as $item)` 

If you don't specify the variable, then you won't have the extra functionality compiled etc, and of course you can name $loop whatever you want:

```
@foreach($items as $item with $my_loop)
    @foreach($items as $item with $my_other_loop)
    @endforeach
@endforeach
```

**Comment by rkgrep** [comment]
:+1: for @sileence idea. I don't think it is a good idea to have a simple loop converted to several method calls by default.

P.S. The [Blade extensions package](http://robin.radic.nl/blade-extensions/) already contains this and much more.

**Comment by helmut** [comment]
Or perhaps `in` makes sense...

`@foreach($items as $item in $loop)`

**Comment by sileence** [comment]
@helmut I like `in` because it's shorter and it makes sense

**Comment by GrahamCampbell** [comment]
Adding this syntax may cause issues with the regex matching ` in ` in other places, like we have other bugs for similar things already.

**Comment by sileence** [comment]
But is it only related to the word `in` @GrahamCampbell ? That's weird. If you point me in the right direction I might be able to help / find a solution.

**Comment by GrahamCampbell** [comment]
Well, we're somewhat limited for a solution due to the representational power of regex.

**Comment by rkgrep** [comment]
Why `in`? That makes no sense! The variable is not IN the `$loop`, it is described by `$loop`. It will also be confusing with `for $var in $array` syntax, that is common for other languages.

**Comment by GrahamCampbell** [comment]
@rkgrep Other languages don't use regular expressions, they are recognised by context free grammars.

**Comment by rkgrep** [comment]
@GrahamCampbell I mean, that the word `in` makes no sense here. If there will be a keyword, it should describe the action.
And I agree with you, that there might have problems with regex here.
What if make a new keyword like `forelse`? E. g. `@loop($array as $var)` `@endloop` or something like this?

**Comment by GrahamCampbell** [comment]
The problem with regex, is that matching brackets is technically impossible, and requires hacks to get it working.

**Comment by GrahamCampbell** [comment]
The bug I was thinking about was our incorrect matching of `or`, and this code would cause the same bug for `in`: https://github.com/laravel/framework/issues/10135.

**Comment by sileence** [comment]
I think using `@loop` could be a good idea as well, that way we don't have to change the default foreach functionality, but I'd to make explicit the declaration of the loop variable.

**Comment by rkgrep** [comment]
What about some special character, that can be escaped properly? E. g. `@foreach($array as $var :: $loop)`. We already use escaping for echo tags.
It is not a common case of using double semicolon with spaces around it also.

**Comment by pun-ky** [comment]
how about adding: $loop->isOdd, $loop->isEven ?:>

**Comment by Modelizer** [comment]
@all What if we do `@foreach($array alias $loop as $key => $value)` where `alias $loop` name will be optional. @sileence even your idea is good 👍

**Comment by CupOfTea696** [comment]
> What if we do @foreach($array alias $loop as $key => $value) where alias $loop name will be optional.

Yes but maybe use `with` rather than alias? Because alias implies that the `$loop` variable will be `$array`.

Lastly, the following would also be useful:

```php
@foreach($items as $item)
     {{ $loop->even }} // iteration is even (index is odd*)
     {{ $loop->odd }} // iteration is odd
     {{ $loop->content }} // contents of $items
@endforeach
```

\* when `$loop->isFirst` is `true`, `$loop->index` would be `0`, while "first" suggests odd rather than even

**Comment by GrahamCampbell** [comment]
@themsaid Thank you very much for this. We're nearly there. Can you please rebase and send a new PR targeting our new 5.3 branch. master is now 5.4. ;)

**Comment by themsaid** [comment]
@GrahamCampbell Alright :)
